keps/sig-storage/1979-object-storage-support/README.md

```diff
@@ -3,703 +3,832 @@
+
+When an admin deletes any of the Class objects, it does not affect existing Buckets as they store a copy of the fields in the class object. If a Bucket  is manually deleted by an admin, then the cluster is considered to be in an invalid state. Manual recovery is possible if data is not already lost.  
+
+# COSI API Reference
+
+## Bucket
+
+Resource to represent a Bucket in OSP. Buckets are cluster-scoped.
+
+```yaml
+Bucket {
+  TypeMeta
+  ObjectMeta
+
+  Spec BucketSpec {
+    // Provisioner is the name of driver associated with this bucket
+    Provisioner string
+
+    // DeletionPolicy is used to specify how COSI should handle deletion of this
+    // bucket. There are 3 possible values:
+    //  - Retain: Indicates that the bucket should not be deleted from the OSP
+    //  - Delete: Indicates that the bucket should be deleted from the OSP
+    //        once all the workloads accessing this bucket are done
+    //  - ForceDelete: Indicates that the bucket and its contents should be
+    //        deleted from the OSP even if workloads are currently accessing
+    //        this bucket
+    DeletionPolicy DeletionPolicy
+
+    // Name of the BucketClass specified in the BucketRequest
+    // +optional
+    BucketClassName  string
+
+    // Name of the BucketClaim that resulted in the creation of this Bucket
+    // +optional
+    BucketClaim corev1.ObjectReference
+
+    // Protocol is the data API this bucket is expected to adhere to.
+    // The possible values for protocol are:
+    // -  S3: Indicates Amazon S3 protocol
+    // -  Azure: Indicates Microsoft Azure BlobStore protocol
+    // -  GCS: Indicates Google Cloud Storage protocol
+    Protocol Protocol
+
+    // Parameters is an opaque map for passing in configuration to a driver
+    // for creating the bucket
+    // +optional
+    Parameters map[string]string
+
+    // BucketID is the unique id of the bucket in the OSP. This field should be
+    // used to specify a bucket that has been created outside of COSI.
+    //
+    // This field will be empty when the Bucket is dynamically provisioned by COSI.
+    // +optional
+    BucketID string
+  }
+
+  Status BucketStatus {
+    // BucketReady is a boolean condition to reflect the successful creation
+    // of a bucket.
+    BucketReady bool
+  }
 }
 ```
 
-## ProvisionerGetInfo
+## BucketClaim
 
-This call is meant to retrieve the unique provisioner Identity. This identity will have to be set in `BucketRequest.Provisioner` field in order to invoke this specific provisioner.
+A claim to create Bucket/bind to existing bucket. BucketClaim is namespace-scoped
 
+```yaml
+BucketClaim {
+  TypeMeta
+  ObjectMeta
+
+  Spec BucketClaimSpec {
+    // Name of the BucketClass
+    BucketClassName string
+  }
```

Comment:
This may have already been discussed - there is no way for the admin to define a prefix, or any portion of a cosi-generated backend bucket name. Early drafts had an optional _prefix_ in the bucketRequest. A driver could be passed a _prefix_ flag that it could apply to all bucket names, or a label could be added to the `Bucket` instance. Just thinking it could be useful for an admin to see all COSI-generated buckets.